### PR TITLE
Initial attempt at documenting MMM01

### PR DESCRIPTION
**Pull request by TheThief**
Mostly based on https://wiki.tauwasser.eu/view/MMM01, along with my own experiences implementing MMM01 support in my emulator

### CODE REVIEW

**Comment by TheThief** [comment]
Where do I need to link to the new MMM01 document to get it added to the table of contents?

**Comment by CasualPokePlayer** [comment]
> 2 KiB (at $A000-$A7FF, then mirrored until $BFFF)
1. Why is 2 KiB in here as an available RAM size to begin with (no commerical game used this). (not targetted specifically at this PR, since that part was here before this).
2. Citation needed on 2 KiB behavior?

**Comment by pinobatch** [comment]
_Family BASIC_ for Family Computer and several Super Famicom/Super NES games have 16 kbit (2048 byte) SRAM, mirrored throughout a 64 kbit (8192 byte) window.

**Comment by TheThief** [comment]
> > 2 KiB (at $A000-$A7FF, then mirrored until $BFFF)
> 
> 1. Why is 2 KiB in here as an available RAM size to begin with (no commerical game used this). (not targetted specifically at this PR, since that part was here before this).
> 2. Citation needed on 2 KiB behavior?

I checked my ROM list: https://docs.google.com/spreadsheets/d/1uEG6wRKeW4fY9th_EFrE23QJaUFd5eXv6Mp4wDBwvUk/edit?usp=sharing

...and 2kB only shows up in mostly PD ROMs (where it's an error, and the flash cartridges used didn't actually contain 2 kB RAM) and _one single_ unlicensed cart (which, incidentally, says it has an MBC5), which may have made the same mistake, and didn't actually contain 2kB.

**Comment by daid** [comment]
Yep, we removed the 2KB RAM from the Pandocs list a while back, the only reference to that value we could find is the (vert old) pandocs, and everything else just copied it from there.

Without any official cart to test the 2KB RAM, there was no way to verify if it wraps, has secretly more RAM or is open-bus for the rest of the address space. (And emulators do all 3 of these)

**Comment by TheThief** [comment]
> Yep, we removed the 2KB RAM from the Pandocs list a while back, the only reference to that value we could find is the (vert old) pandocs, and everything else just copied it from there.
> 
> Without any official cart to test the 2KB RAM, there was no way to verify if it wraps, has secretly more RAM or is open-bus for the rest of the address space. (And emulators do all 3 of these)

From my knowledge of how the carts are _actually wired_, only the first two are possible. And given no official carts have 2 kB and there's no source for the "2 kB" meaning of the RAM size tag in the header, it's most likely these are 8 kB RAM carts

**Comment by TheThief** [comment]
I'll remove it from here while I'm here

**Comment by daid** [comment]
Also, recommend to split this pull request into two, one for the MBC1 changes and one for the MMM01 new mapper. This way it's easier to review and merge.

**Comment by TheThief** [comment]
Will do. They were in together because I came up with much better descriptions of mode 1 while working on the MMM01 docs

**Comment by TheThief** [comment]
MBC1 changes moved to https://github.com/gbdev/pandocs/pull/312

File: src/MMM01.md
Comment (ISSOtm) [inline src/MMM01.md:185]:
```suggestion
2. **Mode register** ($6000) — contains ROM Bank Mask, and Mode
```

File: src/MMM01.md
Comment (ISSOtm) [inline src/MMM01.md:186]:
```suggestion
3. **RAM register** ($4000) — contains Mode Write Disable, and RAM Bank
```

File: src/MMM01.md
Comment (ISSOtm) [inline src/MMM01.md:187]:
```suggestion
4. **RAM Enable register** ($0000) - contains RAM Bank Mask, and Mapping Enable
```

**Review by ISSOtm — CHANGES_REQUESTED** [review]
- Split large paragraphs in the source into more lines to improve editability. Dunno about readability, though.
- Corrected prefixes and hex notation
- Hopefully clarified some wording

We maybe will want more x-refs ([links within the document](https://github.com/gbdev/pandocs#internal-links), for example adding a link to the relevant section whenever the "main ROM banking register" is brought up, which should help clarify. I've marked some, but not all of them.

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md:3]: Lower case "k" is correct for "kilo"

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md:13]: That's not strictly accurate - the extra bits can _only_ be used for game selection. It's the lower MBC1-compatible bits that are maskable to be either for game selection or in-game banking use.

File: src/MMM01.md
Comment (ISSOtm) [reply src/MMM01.md:3]: Except that [this is not kilo, but "kibi"](https://en.wikipedia.org/wiki/Binary_prefix#kibi), for which both "k" and "K" are accepted:

> The first letter of each such prefix is therefore identical to the corresponding SI prefixes, except for "K", which is used interchangeably with "k", whereas in SI, only the lower-case k represents 1000.

[The style guide recommends uppercase K](https://github.com/gbdev/pandocs/wiki/Document-Style#4-units-and-prefixes), too.

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md:13]: Also, the extra bits aren't all in the "correct" registers, so the distinction has to be made between bank number and register

Comment (ISSOtm) [reply src/MMM01.md:13]: Depends on what you consider the "primary" register, I think: are the different bit groups individual registers, which are combined (and transformed) to produce a ROM bank number, or are they just separated views of a single, larger ROM bank register?

I prefer the latter for uniformity, but external opinions are welcome.

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md:3]: Ok that's fair. I'll correct it.

Comment (ISSOtm) [reply src/MMM01.md:3]: I noticed I missed some instances, so please correct these as well. (A case-sensitive `\bkiB\b` regex should take care of it pretty quickly, actually.)

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md]: In the other MBC docs, there's a "registers" header which refers to them by address ranges, and the same is true here - so there's a precedence that "register" refers to the written values. Particularly noteworthy is the MBC1 doc, where the second register can also be used for extending the ROM banking number. This gives a precedence that the bank number is not the bank register.

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md]: Pandocs were originally as much for homebrew development as emulator development - that last line is relevant for homebrew.

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md:26]: The style guide says no $ on addresses

Comment (TheThief) [reply src/MMM01.md:26]: "ROM Bank X0" (with no $) is copied off the MBC1 docs - I've checked and none of the other MBC docs use $ on the bank number in these headers either

Comment (ISSOtm) [reply src/MMM01.md:26]: OK for the addresses, the context is clear enough from the "3FFF"; but I don't think it is for the "X0", so I insist on the $ there. The other docs don't have it because nobody has bothered to fix them yet—Pan Docs is full of such minor errors that have yet to be corrected.

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md:70]: It applies all the time. I'll add that.

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md:177]: There's no dedicated section for it in the MBC1 docs - it's noted throughout. I could link the MBC1 doc maybe?

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md:93]: I've attempted to explain it in the mask section, but here's an example:

$10 written to ROM Bank Low
$10 written to ROM Bank Mask
// entering mapped mode at this point would map actual bank $11 to 4000-7FFF because the unmasked bits are all 0s

So unlike MBC1, the remapping isn't only triggered by a write of $00 to the ROM Bank register. I believe MBC1 is the same (only using a 1, but storing it as 0) but the difference is not observable on an actual MBC1.

File: src/MMM01.md
Comment (ISSOtm) [reply src/MMM01.md:177]: In the meantime, I guess. We should have it in the MBC1 docs, though—perhaps it is in #312 ? I haven't had time to look at it yet.

File: src/MMM01.md
Comment (ISSOtm) [reply src/MMM01.md:93]: Ok, yeah, it's observable, so definitely backed up. Could this rationale be noted? For example:
```suggestion
If all _unmasked_ bits equal 0, the lowest bit, which is always writable, behaves as if set to 1 (as per MBC1 behaviour, mapping bank $00 maps bank $01).
However, the actual value of the register doesn't change, as evidenced if changing the [masking register](...) to un-trigger this behavior.
```

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md:141]: Can you show me an example of what you mean?

Comment (ISSOtm) [reply src/MMM01.md:141]: [This](https://gbdev.io/pandocs/STAT.html#ff41---stat-lcd-status-rw).

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md]: yes

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md]: I've moved it

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md]: I've added an xref for RAM Bank Mask, how can I xref the MBC1 RAM Bank register from another file?

Comment (ISSOtm) [reply src/MMM01.md]: Just use `[MBC RAM Bank register](<#Name of the section>)`, [as explained in the README](https://github.com/gbdev/pandocs#internal-links).

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md]: rewritten this section in a much better way

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md]: Does that work between files?

Comment (ISSOtm) [reply src/MMM01.md]: Yep!

Comment (TheThief) [reply src/MMM01.md]: Yes apparently.

We have the same subsection headings for bank registers in multiple MBC documents - how do we disambiguate which one we're linking to?

Comment (ISSOtm) [reply src/MMM01.md]: ...that's annoying. It'll probably pick up the first section thusly named; we'll have to either change the section names in those pages to become unique, or manually link to the page, anchor included. Maybe I should introduce alternative syntax for this.

Comment (TheThief) [reply src/MMM01.md]: Does _this_ work? https://www.markdownguide.org/extended-syntax/#heading-ids

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md:146]: That's... really ugly.

File: src/MMM01.md
Comment (ISSOtm) [reply src/MMM01.md]: No, it doesn't.

File: src/MMM01.md
Comment (ISSOtm) [reply src/MMM01.md:146]: [We know.](https://github.com/gbdev/pandocs/issues/89)

Comment (TheThief) [reply src/MMM01.md:146]: Can we not keep the diagrams? The table loses the "visual" aspect of it.

I copied the concept from the tauwasser's MMM01 docs and found it very readable there.

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md]: The new version has this:

```
Note: It's suspected the lowest bit of ROM Bank Low (post $00->$01 remapping) still affects the bank mapped to the 4000-7FFF region in "unmapped" mode the same as it does in mapped mode.
Most of the time this would still be a 1, but during game selection it could momentarily go to 0 in between setting the game select bits in [ROM Bank Low](<#Bits 0-4: ROM Bank Low>) and masking them as such in [ROM Bank Mask](<#Bits 1-5: ROM Bank Mask>).
```

Is that better? I've removed a bit of technical detail as the behaviour seems to be unverified.

File: src/MMM01.md
Comment (ISSOtm) [reply src/MMM01.md:146]: If you think the format should be changed (and yes, monospaced text must definitely be removed for this purpose), please open an issue so that we can decide on what replacement is best.

In the meantime, consistency is more important, if only so as to make conversion later on less painful.

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md]: It's not in 312. But that can be added :)

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md:146]: I made one: #318 (I can't seem to label it as "consistency/style")

Comment (ISSOtm) [reply src/MMM01.md:146]: That's because only maintainers can modify labels, of course ;)

File: src/MMM01.md
Comment (ISSOtm) [reply src/MMM01.md]: Okay, but now I don't know what it's trying to imply. That the penultimate 16 KiB bank may be mapped to ROMX in "unmapped" mode under some circumstances?

Comment (TheThief) [reply src/MMM01.md]: Yeah. Basically the suspicion is that the low bit isn't forces to 1 by any "unmapped mode" logic, and just keeps its normal value from the normal banking logic.

This can mean it's 0 in some situations

Comment (TheThief) [reply src/MMM01.md]: Yeah. Basically the suspicion is that the low bit doesn't have any special value in "unmapped mode", and just keeps its normal value from the normal banking logic.

This can mean it's 0 in some situations.

Comment (TheThief) [reply src/MMM01.md]: Which means it ends up mapping bank 1FE to _both_ ROM areas

Comment (ISSOtm) [reply src/MMM01.md]: So what you're saying is that all that "unmapped" mode does, is treat all bank bits *but bit 0* as high?

Comment (TheThief) [reply src/MMM01.md]: It's believed so yes. That and enables write access to the extra MMM01 control bits.

But my only source for that is tanhauser's document, which is confusing at best in places.

Comment (ISSOtm) [reply src/MMM01.md]: Then I would rather word it that way, and explain this caveat at the section about "unmapped" mode, instead of here.

File: src/MMM01.md
Comment (ISSOtm) [reply src/MMM01.md:146]: While we wait for #318 to be resolved (which will entail making some custom markup, so it'll take time), can those be changed to the same template as the rest of the document? They'll be converted in due time, and it's not worth letting this PR rot just for this.

Comment (TheThief) [reply src/MMM01.md:146]: I am currently away

**Comment by ISSOtm** [comment]
Begun another partial review, but the mapping diagrams are basically blocked by #334.

File: src/MMM01.md
Comment (avivace) [inline src/MMM01.md:67]:
```suggestion
When a game is mapped, this area may contain any of the further 16 KiB banks of the game ROM, except for game banks $00, $20, $40, or $60.
```

File: src/MMM01.md
Comment (avivace) [inline src/MMM01.md:70]: We don't use `~` as pseudocode in text. Please review: https://github.com/gbdev/pandocs/wiki/Document-Style

File: src/MMM01.md
Comment (avivace) [inline src/MMM01.md:88]: "suspects" but go in info box reporting "speculation"

File: src/MMM01.md
Comment (avivace) [inline src/MMM01.md:155]: Hex values must be preceded (prefixed). Please see https://github.com/gbdev/pandocs/wiki/Document-Style

**Review by avivace — CHANGES_REQUESTED** [review]

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md:3]: There's supposed to be a number there

File: src/MMM01.md
Comment (TheThief) [reply src/MMM01.md:70]: There are in fact no bitwise operations in the pseudocode section at all, if I'm not blind.

**Pull request merged**
